### PR TITLE
Allow relocatable device code linking in pytorch CUDA extensions

### LINKED ISSUE #57543: Cuda Dynamic Parallelism Support for building custom pytorch library
## 🚀 Feature
I am requesting to add a feature to torch cpp_extenstion to include cuda dynamic parallelism support while building. 

## Motivation

If I am having a global kernel call inside another a global kernel, I need to use ```-rdc=true``` flags for nvcc. That is possible while stand-alone execution of the code. 
I have a cuda script ```query_gpu.cu``` which invokes a ```__global__()``` kernel to perform certain oprations. Inside the kernel, another ```__global__()``` kernel needs to be called. If I build this using torch extension, I get the this error
```undefined symbol: __cudaRegisterLinkedBinary_<number>```

As per [this](https://stackoverflow.com/questions/22115197/dynamic-parallelism-undefined-reference-to-cudaregisterlinkedbinary-linking) stackoverflow answer and [this](https://discuss.pytorch.org/t/custom-cuda-extension-with-dynamic-parallelism/45000) pytorch disussion thread, we need an intermediate step to for device linking. This feature is currently not available in the default setup options for building cuda extensions.

## Pitch
For this, I request to introduce a clear setup for multiple ```nvcc``` command executions, just like one needed here, inside the setup options. Or simple a flag like ```dynamic_parallel=True``` which should be able to take care of everything.

cc @yf225 @glaringlee @ngimel

### PR DESCRIPTION
**Pull request by xwang233**
Close https://github.com/pytorch/pytorch/issues/57543

Doc: check `Relocatable device code linking:` in https://docs-preview.pytorch.org/78225/cpp_extension.html#torch.utils.cpp_extension.CUDAExtension

### CODE REVIEW

**Comment by xwang233** [comment]
cc @ptrblck @nvcastet

File: torch/utils/cpp_extension.py [hunk 1]
```diff
@@ -978,6 +988,19 @@ def CUDAExtension(name, sources, *args, **kwargs):
     Currently open issue for nvcc bug: https://github.com/pytorch/pytorch/issues/69460
     Complete workaround code example: https://github.com/facebookresearch/pytorch3d/commit/cb170ac024a949f1f9614ffe6af1c38d972f7d48
 
+    Relocatable device code linking:
+
+    This extension supports the extra linking step needed for relocatable device code by setting ``dlink`` to ``True``.
+    Relocatable device code libraries can optionally be passed via ``dlink_libraries``.
+    Extra device linking flags (e.g. ``-dlto``) can be passed as a list via the ``nvcc_dlink`` key of ``extra_compile_args``.
```
Comment (ezyang) [inline torch/utils/cpp_extension.py:995]: It's not clear to me from the docs here why I'd want to enable relocatable device code.

Comment (ezyang) [inline torch/utils/cpp_extension.py:995]: It's not clear to me from the docs here why I'd want to enable relocatable device code.

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -2084,6 +2124,15 @@ def sanitize_flags(flags):
         object_file = object_file.replace(" ", "$ ")
         build.append(f'build {object_file}: {rule} {source_file}')
 
+    if cuda_dlink_post_cflags:
+        devlink_out = os.path.join(os.path.dirname(objects[0]), 'dlink.o')
```
Comment (ezyang) [inline torch/utils/cpp_extension.py:2128]: Is there a risk of file name collision here?

Comment (ezyang) [inline torch/utils/cpp_extension.py:2128]: Is there a risk of file name collision here?

**Comment by ezyang** [comment]
can we get a test

**Comment by ezyang** [comment]
can we get a test

File: torch/utils/cpp_extension.py (hunk 2 above)
Comment (xwang233) [reply torch/utils/cpp_extension.py]: fixed

**Review by ezyang — APPROVED** [review]
okey happy to merge when you're ready

**Review by malfet — APPROVED** [review]
LGTM, but isn't it a bit too restrictive? I.e. why Makefile generator would not do?

**Comment by nvcastet** [comment]
> LGTM, but isn't it a bit too restrictive? I.e. why Makefile generator would not do?

@malfet As far as I know, there are only 2 backends in `cpp_extension.py`: ninja and standard distutils.
Unfortunately, the standard distutils is not flexible enough to support an extra device linking step.

**Comment by xwang233** [comment]
Hi @ezyang @malfet , tests are passed and failures seems irrelevant. Can you help merge this PR? Thank you!

**Pull request closed without merging**